messaging/src/main/java/org/axonframework/eventhandling/processors/streaming/token/store/TokenStore.java

```diff
@@ -170,110 +201,112 @@ default void extendClaim(@Nonnull String processorName, int segment) throws Unab
-                     .map(segment -> Segment.computeSegment(segment, allSegments))
-                     .collect(Collectors.toList());
+    @Nonnull
+    default CompletableFuture<List<Segment>> fetchAvailableSegments(@Nonnull String processorName,
+                                                                    @Nullable ProcessingContext context) {
+        return fetchSegments(processorName, context)
+                .thenApply(segments -> Arrays.stream(segments).boxed()
+                                             .map(segment -> Segment.computeSegment(segment, segments))
+                                             .collect(Collectors.toList()));
     }
 
     /**
-     * Returns a unique identifier that uniquely identifies the storage location of the tokens in this store. Two token
-     * store implementations that share state, must return the same identifier. Two token store implementations that
-     * do not share a location, must return a different identifier (or an empty optional if identifiers are not
-     * supported).
+     * Returns a unique identifier that uniquely identifies the storage location of the tokens in this store.
+     * <p>
+     * Two token store implementations that share state <b>must</b> return the same identifier. Two token store
+     * implementations that do not share a location must return a different identifier (or an empty optional if
+     * identifiers are not supported).
      * <p>
-     * Note that this method may require the implementation to consult its underlying storage. Therefore, a Transaction
-     * should be active when this method is called, similarly to invocations like {@link #fetchToken(String, int)},
-     * {@link #fetchSegments(String)}, etc. When no Transaction is active, the behavior is undefined.
+     * Note that this method may require the implementation to consult its underlying storage. Therefore, a transaction
+     * should be active when this method is called, similarly to invocations like
+     * {@link #fetchToken(String, int, ProcessingContext)}, {@link #fetchSegments(String, ProcessingContext)}, etc. When
+     * no transaction is active, the behavior is undefined.
      *
-     * @return an identifier to uniquely identify the storage location of tokens in this TokenStore.
-     * @throws UnableToRetrieveIdentifierException when the implementation was unable to determine its identifier
+     * @return A {@link CompletableFuture} that provides an identifier to uniquely identify the storage location of
+     * tokens in this {@code TokenStore} on completion.
+     * @throws UnableToRetrieveIdentifierException When the implementation was unable to determine its identifier.
      */
-    default Optional<String> retrieveStorageIdentifier() throws UnableToRetrieveIdentifierException {
-        return Optional.empty();
+    @Nonnull
+    default CompletableFuture<Optional<String>> retrieveStorageIdentifier() throws UnableToRetrieveIdentifierException {
```

Comment:
It's the only method we have without `ProcessingContext`, I'm not sure if it's should be an argument here or not... What do you think @smcvb / @nordicthings ?

Reply:
Ah yea, this should also get the context, as now it's the odd one out.